projects/browser-phase-1.md

```diff
@@ -0,0 +1,166 @@
+# Browser Instrumentation (Phase 1) Proposal
+
+## Background and Description
+
+OpenTelemetry is rebooting telemetry for the Browser as a sequence of small, highly focused projects. The first project will focus on creating instrumentation for the browser runtime and a small set of browser libraries, plus the prerequisite API and data modeling work needed to create this instrumentation.
+
+## Current Challenges
+
+OpenTelemetry currently has a NodeJS-focused javascript implementation that is also capable of running in the browser. However, the requirements for a successful browser observability solution differ enough from NodeJS that a specialized solution is required.
+
+* Loading and unloading in the browser are specialized tasks that are very different from booting and shutting down a NodeJS application. Unloading in particular is under strict time and resource constraints.
+* Memory, cpu, and networking resources are severely constrained, especially on mobile devices.
+* A lack of compression and gRPC in browsers
+* Package and dependency sizes
+* All activity occurring in the program relates to a single user instead of many independent transactions. So far this has led to a number of spec developments, such as the need for mutable resources for session management, a better logging model for recording events, and may lead to more browser specific design requirements.
+* Sessions in particular are a very important lifecycle for clients which are not present as a concept in server-side programs.
+* Sessions persist across multiple page loads, another difficult problem to design a solution for that is browser specific.
+* NodeJS has built in facilities for tracking context across async boundaries, but browsers do not have an equivalent concept.
+
+## Goals, Objectives, and Requirements
+
+OpenTelemetry needs a javascript implementation optimized for the browser. Sharing code with NodeJS is helpful when possible. But going forward, we want to prioritize running well in the browser over code reuse with the NodeJS project. If there are places where code reuse with NodeJS is a hard requirement, those places should be identified explicitly.
```

Comment:
This is vague, we should be explicit about the technical direction of API package - which approach will this SIG take?

* The SIG aims to produce a shared OpenTelemetry API package that can be used by both web browsers and Node.js services.
* The SIG aims to produce two separate OpenTelemetry API packages, one for web browser and one for Node.js services.
* The SIG doesn't know the answer yet and would expect to figure this out later.

There are many libraries (e.g., NPM packages) that target both web browsers and Node.js services, these libraries will need an OpenTelemetry API package for instrumentation.

Reply:
My personal opinion is that a separate web and node API is a non-starter for exactly this reason:

> There are many libraries (e.g., NPM packages) that target both web browsers and Node.js services, these libraries will need an OpenTelemetry API package for instrumentation.

I would much prefer to improve the existing API package even if that means a 2.0

Reply:
we can discuss this in the SIG though